geode-core/src/main/java/org/apache/geode/internal/cache/execute/FunctionRemoteContext.java

```diff
@@ -84,6 +85,10 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
       this.bucketArray = BucketSetHelper.fromSet(bucketSet);
     }
     this.isReExecute = DataSerializer.readBoolean(in);
+
+    if (StaticSerialization.getVersionForDataStream(in).isNotOlderThan(KnownVersion.GEODE_1_14_0)) {
```

Comment:
Instead of this, I think you could make this class implement VersionedDataSerializable which might handle writing the principal differently.

Reply:
I tried that but had some upgrade tests that became flaky. I may have done something wrong, but using `VersionedDataSerializable` seemed to not work for this.